zmq.hpp

```diff
@@ -615,8 +620,340 @@ inline void swap(context_t &a, context_t &b) ZMQ_NOTHROW {
+// http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/n4771.pdf
+
+class mutable_buffer
+{
+  public:
+    mutable_buffer() noexcept : _data(nullptr), _size(0) {}
+    mutable_buffer(void *p, size_t n) noexcept : _data(p), _size(n)
+    {
+        assert(p != nullptr || n == 0);
+    }
+
+    void *data() const noexcept { return _data; }
+    size_t size() const noexcept { return _size; }
+    mutable_buffer &operator+=(size_t n) noexcept
+    {
+        // (std::min) is a workaround for when a min macro is defined
+        const auto shift = (std::min)(n, _size);
+        _data = static_cast<char *>(_data) + shift;
+        _size -= shift;
+        return *this;
+    }
+
+  private:
+    void *_data;
+    size_t _size;
+};
+
+inline mutable_buffer operator+(const mutable_buffer &mb, size_t n) noexcept
+{
+    return mutable_buffer(static_cast<char *>(mb.data()) + (std::min)(n, mb.size()),
+                          mb.size() - (std::min)(n, mb.size()));
+}
+inline mutable_buffer operator+(size_t n, const mutable_buffer &mb) noexcept
+{
+    return mb + n;
+}
+
+class const_buffer
+{
+  public:
+    const_buffer() noexcept : _data(nullptr), _size(0) {}
+    const_buffer(const void *p, size_t n) noexcept : _data(p), _size(n) {}
+    const_buffer(const mutable_buffer &mb) noexcept :
+        _data(mb.data()),
+        _size(mb.size())
+    {
+    }
+
+    const void *data() const noexcept { return _data; }
+    size_t size() const noexcept { return _size; }
+    const_buffer &operator+=(size_t n) noexcept
+    {
+        const auto shift = (std::min)(n, _size);
+        _data = static_cast<const char *>(_data) + shift;
+        _size -= shift;
+        return *this;
+    }
+
+  private:
+    const void *_data;
+    size_t _size;
+};
+
+inline const_buffer operator+(const const_buffer &cb, size_t n) noexcept
+{
+    return const_buffer(static_cast<const char *>(cb.data())
+                          + (std::min)(n, cb.size()),
+                        cb.size() - (std::min)(n, cb.size()));
+}
+inline const_buffer operator+(size_t n, const const_buffer &cb) noexcept
+{
+    return cb + n;
+}
+
+// buffer creation
+
+inline mutable_buffer buffer(void* p, size_t n) noexcept
+{
+    return mutable_buffer(p, n);
+}
+inline const_buffer buffer(const void* p, size_t n) noexcept
+{
+    return const_buffer(p, n);
+}
+inline mutable_buffer buffer(const mutable_buffer& mb) noexcept
+{
+    return mb;
+}
+inline mutable_buffer buffer(const mutable_buffer& mb, size_t n) noexcept
+{
+    return mutable_buffer(mb.data(), (std::min)(mb.size(), n));
+}
+inline const_buffer buffer(const const_buffer& cb) noexcept
+{
+    return cb;
+}
+inline const_buffer buffer(const const_buffer& cb, size_t n) noexcept
+{
+    return const_buffer(cb.data(), (std::min)(cb.size(), n));
+}
+
 namespace detail
 {
+// utility functions for containers with data and size
+// data is nullptr if the container is empty
+template<class T> mutable_buffer buffar_mut_ds(T &data) noexcept
```

Comment:
buffar -> buffer here and for several other functions below

I am not a big friend of abbreviations, would you consider changing the abbreviations to the full form?

I assume ds stands for "data and size", so this function "construct a mutable buffer from data and size"? 

It is not completely clear to be what you mean by "containers with data and size". Do you mean a [ContigousContainer](https://en.cppreference.com/w/cpp/named_req/ContiguousContainer). Then it would be better not to use a data member function, but the iterators, such that this only depends on what is guaranteed by the ContiguousContainer concept.

Maybe naming the methods something with "contiguous container" would be more appropriate then.

Reply:
Agree, the name could be improved. Yes, it accepts contiguous containers or views/sequences that have data and size member functions. I don't know if there is a concept for that, but anything that could be used with `std::data` and `std::size`. Maybe `buffer_mutable_contiguous_sequence`.